Mw fixes ksp120

mwerle opened the pull request:
Various fixes to get KIS compiled under KSP 1.2; no detailed testing as yet.

Commit:
KSP 1.2 fix 1 - disambiguate between KSPDev Logger and Unity Logger

Commit:
KSP 1.2 fix 2 - Rename findAttachNode -> FindAttachNode

Commit:
KSP 1.2 fix 3 - explicity check for null

Commit:
KSP fix 4 - remove obsolete class fields

Commit:
KSP 1.2 fix 5 - fix calculating ScienceData transmitted value (?)

Commit:
KSP 1.2 fix 6 - fix finding resources by name

ihsoft commented:
Is any of these changes **required** to run under 1.2? Or they are just cleanups?

UPD. Nevermind. Figured out you mentioned it in the request description. Thanks! I'll use the changes once ready to go with 1.2.

mwerle commented:
The only one I'm not sure about is the fix for the science data transmission value. The original API used no longer exists but not sure if my change calculates the correct value.

mwerle committed:
KSP 1.2 fix 7 - move various object creation from constructors or member initializers to Awake() call.
Also make IconViewer IDisposable and ensure icons are always disposed instead of using the garbage collector since icons contain a GameObject which can only be accessed from the main thread.

mwerle committed:
KSP 1.2 fix 8 - FindModelComponents now returns a List<T> instead of T[].

ihsoft commented on Plugins/Source/KISAddonCursor.cs:
```diff
@@ -89,6 +82,17 @@ sealed class KISAddonCursor : MonoBehaviour {
   }
 
+  void Awake() {
+    hintWindowStyle = new GUIStyle {
```
This is awkward. It's a static member, it should get initialized before anything else is triggered. If it doesn't (e.g. due to KSP cloning) then member must be turned into instance member.

mwerle replied:
Awake () is run as soon as the script is loaded (according to Unity docs) and one of the APIs used on that line threw an error as not being allowed to run in ctors/initializers.

Can change the member to instance but then if there's any chance of the instance getting destructed might also need to manually Dispose of the resource. 

Conversely since there's only once we load the script and set the static even though we may have multiple instances they'll all share the member keeping the pre-1.2 logic. I don't know KIS well enough to know if this is a singleton or not.

ihsoft replied:
Disposing is a problem anyways. Every instantiation of the cursor will create new instance of `GUIStyle`, and old one will be just abandoned. Though, I don't think this kind of resource needs disposing.

That said, converting into instance member should be OK.

ihsoft commented on Plugins/Source/KISAddonPickup.cs:
```diff
@@ -873,10 +873,26 @@ public enum PickupMode {
     draggedPart = null;
   }
 
+  // Sets icon, ensuring any old icon is Disposed
```
Please, in the following changes stick with the code style (2 spaces indentation, no wrapping for opening bracket, etc.). I will change these ones myself once merge is done but it's better to not do second run on the code.

mwerle replied:
Sorry, thought I had done.  Been fighting with monodevelop to stop screwing around with my code formatting.  I'll be more vigilant in future before submitting.

ihsoft commented on Plugins/Source/KISAddonPointer.cs:
```diff
@@ -12,8 +12,8 @@ namespace KIS {
-  public GameObject audioGo = new GameObject();
-  public AudioSource audioBipWrong = new AudioSource();
+  public GameObject audioGo = null;
+  public AudioSource audioBipWrong = null; //new AudioSource();
```
No need to keep commented code.

mwerle replied:
I left it as there was a double-init for audioBipWrong : the initializer was "new AudioSource()" (which raises an error), and Awake() already had the current code for this member. So I'm not sure why that field was being initialized twice; figured the maintainer could then make an executive decision on this.

ihsoft replied:
Comments get lost in the code very quickly since it's unusual for people to review own code once it's submitted :) If mistake is not caught during the review then it turns into bug.

mwerle committed:
Minor code cleanup to adhere to programming standards for this mod.
